daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/DataValue.scala

```diff
@@ -0,0 +1,209 @@
+  type DataValueFloat = DataValue[JFloat, NonNullable with DataValuePrimitiveType]
+  type DataValueByte = DataValue[JByte, NonNullable with DataValuePrimitiveType]
+  type DataValueInt = DataValue[JInt, NonNullable with DataValuePrimitiveType]
+  type DataValueShort = DataValue[JShort, NonNullable with DataValuePrimitiveType]
+  type DataValueDINode = DataValue[DINode, NonNullable with DataValuePrimitiveType]
+  type DataValueUseNilForDefault = DataValue[UseNilForDefaultObj, NonNullable]
+
+  import scala.language.implicitConversions
+
+  @inline implicit def toDataValue(v: JBigDecimal): DataValueBigDecimal = new DataValue(v)
+  @inline implicit def toDataValue(v: DFDLCalendar): DataValueCalendar = new DataValue(v)
+  @inline implicit def toDataValue(v: DFDLDateTime): DataValueDateTime = new DataValue(v)
+  @inline implicit def toDataValue(v: DFDLDate): DataValueDate = new DataValue(v)
+  @inline implicit def toDataValue(v: DFDLTime): DataValueTime = new DataValue(v)
+  @inline implicit def toDataValue(v: Array[Byte]): DataValueByteArray = new DataValue(v)
+  @inline implicit def toDataValue(v: JBoolean): DataValueBool = new DataValue(v)
+  @inline implicit def toDataValue(v: JNumber): DataValueNumber = new DataValue(v)
+  @inline implicit def toDataValue(v: JLong): DataValueLong = new DataValue(v)
+  @inline implicit def toDataValue(v: JDouble): DataValueDouble = new DataValue(v)
+  @inline implicit def toDataValue(v: JBigInt): DataValueBigInt = new DataValue(v)
+  @inline implicit def toDataValue(v: JString): DataValueString = new DataValue(v)
+  @inline implicit def toDataValue(v: URI): DataValueURI = new DataValue(v)
+  @inline implicit def toDataValue(v: JFloat): DataValueFloat = new DataValue(v)
+  @inline implicit def toDataValue(v: JByte): DataValueByte = new DataValue(v)
+  @inline implicit def toDataValue(v: JInt): DataValueInt = new DataValue(v)
+  @inline implicit def toDataValue(v: JShort): DataValueShort = new DataValue(v)
+  @inline implicit def toDataValue(v: DINode): DataValueDINode = new DataValue(v)
+
+  @inline implicit def toDataValue(v: Long): DataValueLong = new DataValue(v: JLong)
+  @inline implicit def toDataValue(v: Double): DataValueDouble = new DataValue(v: JDouble)
+  @inline implicit def toDataValue(v: Boolean): DataValueBool = new DataValue(v: JBoolean)
+  @inline implicit def toDataValue(v: Float): DataValueFloat = new DataValue(v: JFloat)
+  @inline implicit def toDataValue(v: Byte): DataValueByte = new DataValue(v: JByte)
+  @inline implicit def toDataValue(v: Int): DataValueInt = new DataValue(v: JInt)
+  @inline implicit def toDataValue(v: Short): DataValueShort = new DataValue(v: JShort)
+
+  @inline def unsafeFromAnyRef(v: AnyRef) = new DataValue(v)
+  @inline def unsafeFromMaybeAnyRef(v: Maybe[AnyRef]) = {
+    if (v.isDefined) {
+      new DataValue(v.get)
+    } else {
+      NoValue
+    }
+  }
+
+  val NoValue: DataValueEmpty = new DataValue(null)
```

Comment:
Since you have unsafe constructors I suggest there should *never* be a null pointer in the value slot of these things. NoValue should be represented by new DataValue(NoValueIndicator) where NoValueIndicator is some other private sentinel. That way even your unsafe constructor can reject null if passed.

Reply:
I think the question here is what we want `unsafeFromAnyRef(null)` to mean.

In my mind, `null` is AnyRef's version of NoValue, so it makes sense that it gets converted into NoValue. I don't see any reason why our unsafe converter shouldn't allow this. If the caller wants to allow for NoValue, then not using null would mean they need to wrap their conversion in an if statement so they can produce NoValue explicitly.

If they do not want to allow NoValue, then they are responsible for making sure they are providing a legal value, which would including not being null in this case.

It sounds like you want a safeAnyRefToDataValuePrimitive functions. We could write such a function; it would just require a bunch of type checks as well as a null check (and we might also want a simmilar function for our other DataValue types, including safeAnyRefToDataValuePrimitiveNullable).

Reply:
I'll buy into null being AnyRef's version of NoValue.